File: src/main/java/de/twometer/amongus/net/server/ServerSession.java
package de.twometer.amongus.net.server;

import de.twometer.amongus.model.PlayerColor;
import de.twometer.amongus.model.Session;
import de.twometer.amongus.util.RandomUtil;
import de.twometer.amongus.util.Scheduler;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

public class ServerSession extends Session<PlayerConnection> {

    public int tasksFinished;
    public int totalTaskStages;
    public Scheduler.Task votingTask;
    public Scheduler.Task sabotageTask;
    public AtomicInteger fixers = new AtomicInteger(0);

    public final Map<Integer, Integer> votes = new HashMap<>();

    public ServerSession(String gameCode, int host) {
        super(gameCode, host);
    }

    public float getTaskProgress() {
        if (totalTaskStages == 0) return 0;
        else return tasksFinished / (float) totalTaskStages;
    }

    public void findNewHost() {
        if (getPlayers().size() > 0) {
            var newHost = RandomUtil.getRandomItem(getPlayers());
            setHost(newHost.player.id);
        }
    }

    public PlayerColor getRandomFreeColor() {
        var colors = Arrays.stream(PlayerColor.values()).collect(Collectors.toList());
        for (var p : players)
            colors.remove(p.player.color);

        if (colors.size() == 0)
            throw new RuntimeException("Session is full");

        Collections.shuffle(colors);
        return colors.get(0);
    }

    public void broadcast(Object message) {
        for (var player : players)
            player.sendTCP(message);
    }

    public void broadcastExcept(Object message, int except) {
        for (var player : players)
            if (player.player.id != except)
                player.sendTCP(message);
    }

}

File: src/main/kotlin/de/twometer/amogus/server/ServerSession.kt
package de.twometer.amogus.server

import de.twometer.amogus.concurrency.ScheduledTask
import de.twometer.amogus.model.BaseSession
import de.twometer.amogus.model.Location
import de.twometer.amogus.model.PlayerColor
import de.twometer.amogus.model.SabotageType
import java.util.concurrent.atomic.AtomicInteger

class ServerSession(code: String, host: Int) : BaseSession<PlayerClient>(code, host) {

    val surveillanceLock = Any()
    var surveillancePlayers = 0
    val votingTimerLock = Any()
    var votingTimerCompletePlayers = 0
    var tasksCompleted = 0
    var totalTaskStages = 0
    val votes = HashMap<Int, Int>()
    val isFull: Boolean = players.size >= PlayerColor.values().size

    val sabotageLock = Any()
    var currentSabotage: SabotageType? = null
    var sabotageKillTask: ScheduledTask? = null
    var sabotageFixedLocations = HashSet<Location>()
    var sabotageFixingPlayers = 0

    fun broadcast(packet: Any) {
        players.forEach { it.sendTCP(packet) }
    }

    fun broadcastExcept(packet: Any, except: Int) {
        players.filter { it.id != except }.forEach { it.sendTCP(packet) }
    }

    fun randomColor(): PlayerColor {
        return PlayerColor.values().filter { col -> !players.any { it.color == col } }.shuffled().first()
    }

}